src/rgw/rgw_sal_motr.cc

```diff
@@ -2115,52 +2115,77 @@ int MotrObject::get_bucket_dir_ent(const DoutPrefixProvider *dpp, rgw_bucket_dir
 
     rgw_bucket_dir_entry ent_to_check;
 
+    // Check entry in the cache
     if (this->store->get_obj_meta_cache()->get(dpp, this->get_name(), bl) == 0) {
-      iter = bl.cbegin();
-      ent_to_check.decode(iter);
-      if (ent_to_check.is_current()) {
+        iter = bl.cbegin();
+        ent_to_check.decode(iter);
         ent = ent_to_check;
         rc = 0;
         goto out;
-      }
-    }
-
-    ldpp_dout(dpp, 20) <<__func__<< ": versioned bucket!" << dendl;
-    keys[0] = this->get_name();
-    rc = store->next_query_by_name(bucket_index_iname, keys, vals);
-    if (rc < 0) {
-      ldpp_dout(dpp, 0) << __func__ << "ERROR: NEXT query failed. " << rc << dendl;
-      return rc;
     }
 
-    rc = -ENOENT;
-    for (const auto& bl: vals) {
-      if (bl.length() == 0)
-        break;
+    if(this->have_instance())
+    {
+      // TODO : Handle null version-id scenarios
 
-      iter = bl.cbegin();
-      ent_to_check.decode(iter);
-      if (ent_to_check.is_current()) {
-        ldpp_dout(dpp, 20) <<__func__<< ": found current version!" << dendl;
+      // Cache miss.
+      rc = this->store->do_idx_op_by_name(bucket_index_iname,
+                          M0_IC_GET, this->get_key().to_str(), bl);
+      if(rc<0)
+        return rc;
+      else
+      {
+        iter = bl.cbegin();
+        ent_to_check.decode(iter);
         ent = ent_to_check;
         rc = 0;
-
+        // Put into the cache
         this->store->get_obj_meta_cache()->put(dpp, this->get_name(), bl);
-
-        break;
+        goto out;
       }
     }
-  } else {
-    if (this->store->get_obj_meta_cache()->get(dpp, this->get_key().to_str(), bl)) {
+    else
+    {  // Version-id instance is empty
+       // Cache miss.
+        keys[0] = this->get_name();
+        rc = store->next_query_by_name(bucket_index_iname, keys, vals);
```

Comment:
As I understood, if an instance is empty then the key should be "object_name" else it should be "object_name[instance_name]",
If this is the only diff we don't need to list all the 1000 keys starting from keys[0]. We can use the earlier do_idx_op_by_name() to get the same key.

Reply:
`As I understood, if an instance is empty then the key should be "object_name" else it should be "object_name[instance_name]"`
This is correct.

But, if we pass the key as "object_name" to the do_idx_op_by_name(), it would return the correct object entry but, with empty version-id. Hence, version-id field might be missing in the response. 

For the solution, I have passed the prefix parameter to next_query_by_name() function, so that it will return only the entries of given keys instead of max (1000) keys. This will then return the current object version entry in the response.